qiskit/qasm3/exporter.py

```diff
@@ -361,124 +419,192 @@ def build_definition(self, instruction, builder):
-            params.append(Identifier(param_name))
-        params += [Identifier(param.name) for param in gate.definition.parameters]
-
-        self.circuit_ctx.append(gate.definition)
-        qargList = []
-        for qreg in gate.definition.qregs:
-            for qubit in qreg:
-                qreg, idx = self.find_bit(qubit)
-                qubit_name = f"{qreg.name}_{idx}"
-                qargList.append(Identifier(qubit_name))
-        self.circuit_ctx.pop()
-
-        return QuantumGateSignature(Identifier(name), qargList, params or None)
-
-    def build_inputs(self):
-        """Builds a list of Inputs"""
-        ret = []
-        for param in self.circuit_ctx[-1].parameters:
-            ret.append(IO(IOModifier.input, Identifier("float[32]"), Identifier(param.name)))
-        return ret
+        for num in range(len(gate.params) - len(definition.parameters)):
+            param_name = f"{self.gate_parameter_prefix}_{num}"
+            params.append(ast.Identifier(param_name))
+        params += [ast.Identifier(param.name) for param in definition.parameters]
+        quantum_arguments = [
+            ast.Identifier(f"{self.gate_qubit_prefix}_{n_qubit}")
+            for n_qubit in range(len(definition.qubits))
+        ]
+        return ast.QuantumGateSignature(ast.Identifier(name), quantum_arguments, params or None)
+
+    def build_variable_declarations(self):
+        """Builds lists of the input, output and standard variables used in this program."""
+        inputs, outputs, variables = [], [], []
+        global_scope = self.global_scope(assert_=True).circuit
+        for parameter in global_scope.parameters:
+            declaration = _infer_variable_declaration(global_scope, parameter)
+            if declaration is None:
+                continue
+            if isinstance(declaration, ast.IODeclaration):
+                if declaration.modifier is ast.IOModifier.input:
+                    inputs.append(declaration)
+                else:
+                    outputs.append(declaration)
+            else:
+                variables.append(declaration)
+        return inputs, outputs, variables
 
-    def build_bitdeclarations(self):
-        """Builds a list of BitDeclarations"""
-        ret = []
-        for creg in self.circuit_ctx[-1].cregs:
-            ret.append(BitDeclaration(Identifier(creg.name), Designator(Integer(creg.size))))
-        return ret
+    @property
+    def base_classical_register_name(self):
+        """The base register name"""
+        name = "_all_clbits" if self.alias_classical_registers else "_loose_clbits"
+        if name in self.global_namespace._data:
+            raise NotImplementedError  # TODO choose a different name if there is a name collision
+        return name
 
     @property
-    def base_register_name(self):
+    def base_quantum_register_name(self):
         """The base register name"""
-        name = "_q"
+        name = "_all_qubits"
```

Comment:
Is this worth changing? The register is going to be in every output and it's kinda long-named. maybe @kdk has an opinion?

Reply:
Generated code isn't meant to be read, really, it's meant to be an interchange format for something else.  Given that we don't have any method for handling a naming collision, it's better to be more verbose to minimise the chance of a collision.

I was also having the problem that with everything being single characters, it was hard to spot the problems in diffs when stuff was going wrong, and things like that.

Reply:
We also had `_q[0]` and `q_0` depending on whether we were in a gate or the main programme, which weren't so easy to immediately tell apart at a glance.